drake/systems/plants/RigidBodySystem.cpp

```diff
@@ -32,31 +37,38 @@ RigidBodySystem::StateVector<double> RigidBodySystem::dynamics(const double& t,
   auto H = tree->massMatrix(kinsol);
   Eigen::MatrixXd H_and_neg_JT = H;
 
-  { // loop through rigid body force elements and accumulate f_ext
+  VectorXd C = tree->dynamicsBiasTerm(kinsol,f_ext);
+  if (num_actuators > 0) C -= tree->B*u.topRows(num_actuators);
+
+  { // loop through rigid body force elements
 
     // todo: distinguish between AppliedForce and ConstraintForce
-    // todo: have AppliedForce output tau (instead of f_ext).  it's more direct, and just as easy to compute.
-
-    int u_index = 0;
-    const NullVector<double> force_state;  // todo:  will have to handle this case
-    for (auto const & prop : props) {
-      RigidBodyFrame* frame = prop->getFrame();
-      RigidBody* body = frame->body.get();
-      int num_inputs = 1;  // todo: generalize this
-      RigidBodyPropellor::InputVector<double> u_i(u.middleRows(u_index,num_inputs));
-      // todo: push the frame to body transform into the dynamicsBias method?
-      Matrix<double,6,1> f_ext_i = transformSpatialForce(frame->transform_to_body,prop->output(t,force_state,u_i,kinsol));
-      if (f_ext.find(body)==f_ext.end()) f_ext[body] = f_ext_i;
-      else f_ext[body] = f_ext[body]+f_ext_i;
+
+    size_t u_index = 0;
+    for (auto const & f : force_elements) {
+      size_t num_inputs = f->getNumInputs();
+      VectorXd force_input(u.middleRows(u_index,num_inputs));
+      C -= f->output(t,force_input,kinsol);
       u_index += num_inputs;
     }
   }
 
-  VectorXd C = tree->dynamicsBiasTerm(kinsol,f_ext);
-  if (num_actuators > 0) C -= tree->B*u.topRows(num_actuators);
+  { // apply joint limit forces
+    for (auto const& b : tree->bodies) {
+      if (!b->hasParent()) continue;
+      auto const& joint = b->getJoint();
+      if (joint.getNumPositions()==1 && joint.getNumVelocities()==1) {  // taking advantage of only single-axis joints having joint limits makes things easier/faster here
+        double qmin = joint.getJointLimitMin()(0), qmax = joint.getJointLimitMax()(0);
+        // tau = k*(qlimit-q) - b(qdot)
+        if (q(b->position_num_start)<qmin)
+          C(b->velocity_num_start) -= penetration_stiffness*(qmin-q(b->position_num_start)) - penetration_damping*v(b->velocity_num_start);
+        if (q(b->position_num_start)>qmax)
```

Comment:
`else if` here? Out of curiosity, if `qmin` is very close to `qmax`, how shall we guarantee that the joint limit force does not cause the joint to oscillate? Can we verify a certain choice of stiffness and damping should always work for some joint limit and time step?